File: app/src/main/java/com/watermelon/UI/Watchlist/WatchlistViewModel.java
package com.watermelon.UI.Watchlist;

import android.util.Pair;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.watermelon.Models.TvSeriesFull;
import com.watermelon.UI.WatermelonActivity;
import com.watermelon.domain.usecase.ChangeEpisodeWatchedFlagUseCase;
import com.watermelon.domain.usecase.GetWatchlistUseCase;
import com.watermelon.UI.framework.common.UseCaseHandler;

import java.util.List;

public class WatchlistViewModel extends ViewModel {

    private UseCaseHandler useCaseHandler;
    private GetWatchlistUseCase getWatchlistUseCase;

    private ChangeEpisodeWatchedFlagUseCase changeEpisodeWatchedFlagUseCase;
    private MutableLiveData<List<TvSeriesFull>> watchlistList;

    public WatchlistViewModel(UseCaseHandler useCaseHandler, GetWatchlistUseCase getWatchlistUseCase, ChangeEpisodeWatchedFlagUseCase changeEpisodeWatchedFlagUseCase) {

        this.useCaseHandler  = useCaseHandler;
        this.getWatchlistUseCase = getWatchlistUseCase;
        this.changeEpisodeWatchedFlagUseCase = changeEpisodeWatchedFlagUseCase;
        this.watchlistList = new MutableLiveData<>();
    }

    public void loadWatchlist() {
        useCaseHandler.execute(getWatchlistUseCase, new GetWatchlistUseCase.RequestValues(WatermelonActivity.TVSERIES_WATCHED_FLAG_YES), new UseCaseHandler.UseCaseCallback<GetWatchlistUseCase.ResponseValue>() {
            @Override
            public void onSuccess(GetWatchlistUseCase.ResponseValue response) {
                updateWatchlist(response.getWatchlistList());
            }

            @Override
            public void onError(Exception exception) {
                throw new RuntimeException(exception);
            }
        });
    }

    private void updateWatchlist(List<TvSeriesFull> list) {
        watchlistList.setValue(list);
    }

    LiveData<List<TvSeriesFull>> getWatchlistListObservable() {
        return watchlistList;
    }

    void changeEpisodeWatchedFlag(Pair<Integer, Boolean> params) {
        useCaseHandler.execute(changeEpisodeWatchedFlagUseCase, new ChangeEpisodeWatchedFlagUseCase.RequestValues(params), new UseCaseHandler.UseCaseCallback<ChangeEpisodeWatchedFlagUseCase.ResponseValue>() {
            @Override
            public void onSuccess(ChangeEpisodeWatchedFlagUseCase.ResponseValue response) {
                loadWatchlist();
            }

            @Override
            public void onError(Exception exception) {

            }
        });
    }


}

File: app/src/main/java/com/watermelon/UI/WatchlistCompose/WatchlistViewModel.kt
package com.watermelon.UI.WatchlistCompose

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.watermelon.Models.TvSeriesFull
import com.watermelon.UI.framework.common.UseCaseHandler
import com.watermelon.domain.usecase.GetWatchlistUseCase
import java.lang.Exception

class WatchlistViewModel(
        private val useCaseHandler: UseCaseHandler,
        private val getWatchlistUseCase: GetWatchlistUseCase
) : ViewModel() {
    private val _uiState = MutableLiveData<UiState>()
    val uiState: LiveData<UiState> = _uiState

    fun loadData() {
        useCaseHandler.execute(getWatchlistUseCase, GetWatchlistUseCase.RequestValues(true), object : UseCaseHandler.UseCaseCallback<GetWatchlistUseCase.ResponseValue> {
            override fun onSuccess(response: GetWatchlistUseCase.ResponseValue?) {
                if (response != null) {
                    _uiState.value = UiState(response.watchlistList)
                }
            }

            override fun onError(exception: Exception?) {
                _uiState.value = UiState(emptyList())
            }
        })
    }
}

data class UiState(val data: List<TvSeriesFull>)